hyper/http20/stream.py

```diff
@@ -109,11 +109,22 @@ def __init__(self,
         self._encoder = header_encoder
         self._decoder = header_decoder
 
-    def add_header(self, name, value):
+    def add_header(self, name, value, replace=False):
         """
         Adds a single HTTP header to the headers to be sent on the request.
         """
-        self.headers.append((name.lower(), value))
+        name = name.lower()
+
+        if replace:
+            try:
+                idx = [n for n, _ in self.headers].index(name)
```

Comment:
So, this is a really unfortunate n^2 algorithm that doesn't need to be. Instead, lets change `self.headers` to be an instance of the `HTTPHeaderMap`. This will probably break some tests that do equality tests, but we can fix this by just getting `list(x.headers.items())` for anywhere that that happens.

You may want to add a `replace` method to the `HTTPHeaderMap` that first does `del` then inserts the item into `self` but that should be simple enough. This should avoid some of the duplicate code here.